lib/Sema/TypeCheckCompilerEvaluable.cpp

```diff
@@ -26,26 +26,205 @@ using namespace swift;
+      return {false, E};
+    }
+
+    switch (E->getKind()) {
+    #define ALWAYS_ALLOWED(ID)                                                 \
+      case ExprKind::ID:                                                       \
+        return {true, E};
+    #define SOMETIMES_ALLOWED(ID)                                              \
+      case ExprKind::ID:                                                       \
+         return checkExpr##ID(cast<ID##Expr>(E));
+
+    ALWAYS_ALLOWED(NilLiteral)
+    ALWAYS_ALLOWED(IntegerLiteral)
+    ALWAYS_ALLOWED(BooleanLiteral)
+    ALWAYS_ALLOWED(MagicIdentifierLiteral)
+    ALWAYS_ALLOWED(DiscardAssignment)
+    SOMETIMES_ALLOWED(DeclRef)
+    ALWAYS_ALLOWED(Type)
+    SOMETIMES_ALLOWED(OtherConstructorDeclRef)
+    ALWAYS_ALLOWED(DotSyntaxBaseIgnored)
+    ALWAYS_ALLOWED(MemberRef)
+    ALWAYS_ALLOWED(Paren)
+    ALWAYS_ALLOWED(DotSelf)
+    ALWAYS_ALLOWED(Try)
+    ALWAYS_ALLOWED(ForceTry)
+    ALWAYS_ALLOWED(OptionalTry)
+    ALWAYS_ALLOWED(Tuple)
+    ALWAYS_ALLOWED(Subscript)
+    ALWAYS_ALLOWED(TupleElement)
+    ALWAYS_ALLOWED(CaptureList)
+    ALWAYS_ALLOWED(Closure)
+    ALWAYS_ALLOWED(AutoClosure)
+    ALWAYS_ALLOWED(InOut)
+    ALWAYS_ALLOWED(DynamicType)
+    ALWAYS_ALLOWED(RebindSelfInConstructor)
+    ALWAYS_ALLOWED(BindOptional)
+    ALWAYS_ALLOWED(OptionalEvaluation)
+    ALWAYS_ALLOWED(ForceValue)
+    SOMETIMES_ALLOWED(Call)
+    ALWAYS_ALLOWED(PrefixUnary)
+    ALWAYS_ALLOWED(PostfixUnary)
+    ALWAYS_ALLOWED(Binary)
+    ALWAYS_ALLOWED(DotSyntaxCall)
+    ALWAYS_ALLOWED(ConstructorRefCall)
+    ALWAYS_ALLOWED(Load)
+    ALWAYS_ALLOWED(TupleShuffle)
+    ALWAYS_ALLOWED(InjectIntoOptional)
+    ALWAYS_ALLOWED(Coerce)
+    ALWAYS_ALLOWED(If)
+    ALWAYS_ALLOWED(Assign)
+    ALWAYS_ALLOWED(CodeCompletion)
+    ALWAYS_ALLOWED(EditorPlaceholder)
+
+    // Allow all errors and unchecked expressions so that we don't put errors
+    // on top of expressions that alrady have errors.
+    ALWAYS_ALLOWED(Error)
+    ALWAYS_ALLOWED(UnresolvedTypeConversion)
+    #define UNCHECKED_EXPR(ID, PARENT) ALWAYS_ALLOWED(ID)
+    #include "swift/AST/ExprNodes.def"
+
+    default:
+      TC.diagnose(E->getStartLoc(),
+                  diag::compiler_evaluable_forbidden_expression)
+          .highlight(E->getSourceRange());
+      CompilerEvaluable = false;
+      return {false, E};
+
+    #undef ALWAYS_ALLOWED
+    #undef SOMETIMES_ALLOWED
+    }
+  }
+
+  std::pair<bool, Expr *> checkExprCall(CallExpr *call) {
+    // Allow calls to some stdlib assertion functions without walking them
+    // further, because the calls do currently-forbidden things. (They use
+    // Strings and they call functions imported from C).
+    if (auto *calleeRef = dyn_cast<DeclRefExpr>(call->getDirectCallee()))
+      if (auto *callee = dyn_cast<AbstractFunctionDecl>(calleeRef->getDecl()))
+        if (callee->isChildContextOf(TC.Context.TheStdlibModule) &&
+            (callee->getNameStr() == "_precondition" ||
```

Comment:
Please file a JIRA tracking this so we can discuss it when the implementation is further baked.  I don't want these to fall off the radar.  I think they should be handleable with the   @compilerEvaluable(magicBehavior) feature that we'll need anyway.

Reply:
done: https://bugs.swift.org/browse/SR-8035